### PR TITLE
Bug fixes with variable handling in `LossScaleOptimizer`.

### PR DESCRIPTION
**Pull request by hertschuh**
## overwrite_with_gradient

`overwrite_with_gradient` would be ineffective on JAX in real-world conditions, i.e. within `model.fit`.
    
This is because in the training loop, `stateless_apply` is passed `trainable_variables` as arrays containing the values of the trainable variables, not the variables themselves. Instead, we have to inspect the variables.

## apply with gradients only

`apply(grads)` without the `trainable_variables` argument passed in would not apply anything.
    
This is because the code uses `self._trainable_variables`. But this was an empty array for `LossScaleOptimizer`. This was fixed by adding `super().build(...)`.

## unsupported optimizer arguments
    
Also fail when other arguments from the base optimizer are passed to `LossScaleOptimizer.__init__` since they are not actually supported. They are also no longer returned by `get_config`.

### CODE REVIEW

**Comment by hertschuh** [comment]
@cantonios can you review?

File: keras/src/optimizers/loss_scale_optimizer.py [hunk 1]
```diff
@@ -136,7 +136,9 @@ def increment():
                 g
                 if g is None or self._overwrite_variable_with_gradient(v)
                 else ops.divide(g, scale)
-                for g, v in zip(grads, trainable_variables)
+                for g, v in zip(
+                    grads, self.inner_optimizer._trainable_variables
```
Comment (cantonios) [inline keras/src/optimizers/loss_scale_optimizer.py]: Summarizing from our brief in-person discussion:
I don't think we should pull directly from the inner optimizer here.  Instead, we should ensure the loss-scale optimizer has the correct `_trainable_variables` (maybe from calling it's `super().build(...)`).

File: keras/src/optimizers/loss_scale_optimizer.py [hunk 2]
```diff
@@ -177,7 +179,7 @@ def apply(self, grads, trainable_variables=None):
     def _stateful_handle_finite_grads(self, grads, trainable_variables):
         scale = self.dynamic_scale
         # Unscale gradients.
-        tvs = trainable_variables or self._trainable_variables
+        tvs = trainable_variables or self.inner_optimizer._trainable_variables
```
[the remark this answers is not in the document]
Comment (cantonios) [reply keras/src/optimizers/loss_scale_optimizer.py]: ... which would fix this as well.

File: keras/src/optimizers/loss_scale_optimizer_test.py [hunk 3]
```diff
@@ -40,7 +53,9 @@ def test_finite_step(self, stateless):
         if stateless:
             optimizer.build(vars)
             vars, _ = optimizer.stateless_apply(
-                optimizer.variables, grads, vars
+                [v.value for v in optimizer.variables],
```
Comment (cantonios) [inline keras/src/optimizers/loss_scale_optimizer_test.py:56]: I guess this is because of that JAX deprecation issue?

Comment (cantonios) [reply keras/src/optimizers/loss_scale_optimizer_test.py:56]: How prevalent is this pattern across keras?  If we can no longer rely on the `__jax_array__` built-in to auto convert, maybe there should be a helper function like `optimizer.variable_values` that unwraps the variables?

File: keras/src/optimizers/loss_scale_optimizer.py (hunk 1 above)
Comment (hertschuh) [reply keras/src/optimizers/loss_scale_optimizer.py]: Will call `super().build(...)` and use `self._trainable_variables`.

File: keras/src/optimizers/loss_scale_optimizer.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (hertschuh) [reply keras/src/optimizers/loss_scale_optimizer.py]: Will revert since `super().build(...)` will take care of this.

File: keras/src/optimizers/loss_scale_optimizer_test.py (hunk 3 above)
Comment (hertschuh) [reply keras/src/optimizers/loss_scale_optimizer_test.py:56]: I uncovered this as part of the `__jax_array__` deprecation. See https://github.com/keras-team/keras/pull/21702

However, I discovered in the process that we're testing `stateless_apply` the wrong way. `stateless_apply` takes tensors, not variables (unlike the names suggest). So this change of the unit tests is correct, and is independent of the `__jax_array__` deprecation.

This pattern of unwrapping variables was only added in this test file (see https://github.com/keras-team/keras/pull/21702 ). This pattern does exists to some extend, but in the JAX trainer, not the optimizers.

File: keras/src/optimizers/loss_scale_optimizer.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (hertschuh) [reply keras/src/optimizers/loss_scale_optimizer.py]: Fixed

File: keras/src/optimizers/loss_scale_optimizer.py [hunk 4]
```diff
@@ -136,7 +136,9 @@ def increment():
                 g
                 if g is None or self._overwrite_variable_with_gradient(v)
                 else ops.divide(g, scale)
-                for g, v in zip(grads, trainable_variables)
+                for g, v in zip(
+                    grads, self.inner_optimizer._trainable_variables
+                )
```
[the remark this answers is not in the document]
Comment (hertschuh) [reply keras/src/optimizers/loss_scale_optimizer.py]: Fixed

File: keras/src/optimizers/loss_scale_optimizer.py (hunk 1 above)
Comment (hertschuh) [reply keras/src/optimizers/loss_scale_optimizer.py]: Fixed

File: keras/src/optimizers/loss_scale_optimizer_test.py (hunk 3 above)
Comment (cantonios) [reply keras/src/optimizers/loss_scale_optimizer_test.py:56]: Thanks, this all looks better.
